feat(console): render Cycle schema to PHP file (+roles filter, overwrite)

petrdobr opened the pull request:
- New option to write Cycle ORM schema to a PHP file for `cycle:render`
- Filter by roles to limit the exported subset
- Overwrite flag to safely replace existing file

<!--- Note to EXTERNAL Contributors -->
<!-- Thanks for opening a PR!
If it is a significant code change, please **make sure there is an open issue** for this.
We work best with you when we have accepted the idea first before you code. -->

<!--- For ALL Contributors 👇 -->

## What was changed

- Extended the existing `cycle:render` console command:
  - `-o, --output=/path/to/schema.php` — writes rendered schema to a file (supported **only for** `format=php`).
  - `--overwrite` — allows replacing an existing output file.
  - `-r, --role=<role> [--role=<role> ...]` — filters the rendered schema by roles; accepts multiple flags and comma-separated values (e.g. `--role=auth,rbac --role=default`).
- STDOUT behavior for all formats is preserved. If `--output` is not provided, output goes to STDOUT as before.
- For non-`php` formats with `--output`, the command fails with an explicit error.
- Safe write: content is written to a temporary file in the target directory and then atomically moved to the destination; parent directories are created as needed.

**Examples**
```bash
# to stdout (unchanged)
php app.php cycle:render php

# to file
php app.php cycle:render php --output=var/schema.php

# filter by roles (comma-separated, or multiple --role flags)
php app.php cycle:render php --output=var/schema.auth.php --role=auth,rbac
php app.php cycle:render php --output=var/schema.default.php --role=default --overwrite
```
### Edge cases & messages
- Unknown roles are reported as a warning and ignored.
- If the role filter yields no entries:
    - with `--output` → prints “Nothing to write.” and exits successfully;
    - without `--output` → prints an empty line to STDOUT.
- If the destination file already exists without `--overwrite`, the command fails with an explicit message.
- Directory creation and file write failures are reported with descriptive errors.

## Why?
- A deterministic PHP artifact of the schema is useful for audits, code reviews, caching in CI and reproducible builds.
- Role-based filtering keeps artifacts minimal and context-specific (e.g., exporting only auth-related roles).
- Preserving STDOUT behavior keeps the command backward compatible, while `--output` enables file-based workflows.

## Checklist

- Closes #29
- Tested
    - [x] Tested manually
    - [x] Unit tests added

## Documentation <!--- Remove if not needed -->
**New options:**
- `-o, --output=PATH` — write rendered schema to a PHP file (supported **only** when `format=php`).
- `--overwrite` — allow replacing an existing output file.
- `-r, --role=ROLE` — filter exported schema by roles (for any format).  
  Accepts multiple flags and comma-separated values inside a flag (e.g. `--role=auth,rbac --role=default`).  
  Role matching is case-insensitive; unknown roles are ignored with a warning.

**Behavior & edge cases:**
- Without `--output` the command writes to **STDOUT** (unchanged behavior).
- If `--output` is used with a non-`php` format → the command fails with an explicit error.
- If filtering results in an empty schema:
  - with `--output` → prints “Nothing to write.” and exits successfully;
  - without `--output` → prints an empty line to STDOUT.
- Safe write: schema is written to a temp file in the target dir and then atomically moved; parent dirs are created if missing.

**Examples**
```bash
# stdout (unchanged)
php app.php cycle:render php

# write to file
php app.php cycle:render php --output=var/schema.php

# filter by roles (comma-separated or multiple flags)
php app.php cycle:render php --output=var/schema.auth.php --role=auth,rbac
php app.php cycle:render php --output=var/schema.default.php --role=default --overwrite
```

petrdobr committed:
feat(console): render Cycle schema to PHP file with overwrite and role filtering

- New console command to export Cycle ORM schema into a PHP file
- Filter by roles to limit the exported subset
- Overwrite flag to safely replace existing file

roxblnfk commented on src/Console/Command/CycleOrm/RenderCommand.php:
Could you refactor the command to attributes instead of constants?
See https://spiral.dev/docs/console-commands/

petrdobr replied:
Sure! Will update soon

petrdobr replied:
Done

roxblnfk commented on src/Console/Command/CycleOrm/RenderCommand.php:
```diff
@@ -9,35 +9,175 @@
+    private ?string $outputPath = null;
+
+    #[Option(name: 'overwrite', description: 'Overwrite existing output')]
+    private bool $overwrite = false;
```
I think we don't need this parameter because it should be always overwritten

roxblnfk commented on src/Console/Command/CycleOrm/RenderCommand.php:
```diff
@@ -9,35 +9,175 @@
-    protected const SIGNATURE = 'cycle:render {format=color : Output format}';
-    protected const DESCRIPTION = 'Render available CycleORM schemas';
+    #[Argument(name: 'format', description: 'The format of the output')]
+    private string $format;
```
Let's make `format` as an option, and `role` as an argument.
Also need to set the default value. It can be `null` that means `plain` or `color` depending on terminal capabilities

petrdobr replied:
Got it

petrdobr replied:
And if null is passed for role, for example `php app.php cycle:render` without role argument, then print/save all roles?

roxblnfk replied:
yes
full schema

roxblnfk commented on src/Console/Command/CycleOrm/RenderCommand.php:
```diff
@@ -9,35 +9,175 @@
+    name: 'cycle:render',
+    description: 'Render available CycleORM schemas')
+]
 final class RenderCommand extends AbstractCommand
```
Need to add some docs about how to use. Like:

`php app.php cycle:render user,post,comment`

`php app.php cycle:render --format=php --output=cycle-schema.php`

roxblnfk replied:
I meant PHP comments [like this](https://github.com/php-internal/dload/blob/81578209dfb3ce99caa7df67eeb4ae9715d9585f/src/Command/Get.php#L22-L47) 😃

roxblnfk commented on src/Console/Command/CycleOrm/RenderCommand.php:
```diff
@@ -9,35 +9,175 @@
+            if ($this->format !== 'php') {
+                $this->error('The --output option is currently supported only with format=php.');
+                return self::FAILURE;
+            }
```
Can't we support all the formats?

petrdobr replied:
We can, will do. I thought we only need files for php

roxblnfk commented on src/Console/Command/CycleOrm/RenderCommand.php:
```diff
@@ -9,35 +9,175 @@
+                @\unlink($tmp);
+                $this->error(\sprintf('Failed to write schema to temp file in "%s".', $tmpDir));
+                return self::FAILURE;
+            }
```
Why do we need to use temporary file? Can't we write the target file directly?

petrdobr replied:
I had hot debates with chatGPT and he convinced me that this approach is advantageous. The idea is to write a file "atomically" as a whole and prevent corrupted files if something goes wrong during the writing process. The temporary file will not be saved to the disk if an error occurs.
But I'm not sure if this is an overcomplication or a good idea.

roxblnfk replied:
I think using of file_put_contents is good enough to not to use tmp files :)

roxblnfk commented:
And eah, please don't overwrite reviewed commits via `force pushe`

petrdobr commented:
> And eah, please don't overwrite reviewed commits via `force pushe`

Understood, sorry for rewriting them earlier and thanks for the advice :)

petrdobr committed:
feat(console): improve cycle:render (roles arg, format, file output)

- Make roles an optional argument (comma-separated). If omitted export full schema.
- Add --format=php|mermaid|color|plain; default is
  color for ANSI-capable terminals, otherwise plain.
- Allow --output for all formats; automatically overwrite existing files.
- Update command description/help with usage examples.

roxblnfk committed:
refactor: Add command comment block;
simplify and apply code style

roxblnfk committed:
refactor: Simplify command logic

roxblnfk commented:
Good feature, thanks
